compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/Utils.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.common.lower.loops
+
+import org.jetbrains.kotlin.ir.expressions.IrConst
+import org.jetbrains.kotlin.ir.expressions.IrConstKind
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrGetValue
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.toKotlinType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+internal fun IrExpression.castIfNecessary(targetType: IrType, numberCastFunctionName: Name): IrExpression {
+    return if (type.toKotlinType() == targetType.toKotlinType()) {
+        this
+    } else {
+        val function = type.getClass()!!.functions.first { it.name == numberCastFunctionName }
+        IrCallImpl(startOffset, endOffset, function.returnType, function.symbol)
+            .apply { dispatchReceiver = this@castIfNecessary }
+    }
+}
+
+/** Return the negated value if the expression is const, otherwise call unaryMinus(). */
+internal fun IrExpression.negate(): IrExpression {
+    val value = (this as? IrConst<*>)?.value as? Number
+    return when (value) {
+        is Int -> IrConstImpl(startOffset, endOffset, type, IrConstKind.Int, -value)
+        is Long -> IrConstImpl(startOffset, endOffset, type, IrConstKind.Long, -value)
+        else -> {
+            val unaryMinusFun = type.getClass()!!.functions.first { it.name == OperatorNameConventions.UNARY_MINUS }
+            IrCallImpl(startOffset, endOffset, type, unaryMinusFun.symbol, unaryMinusFun.descriptor).apply {
+                dispatchReceiver = this@negate
+            }
+        }
+    }
+}
+
+/** Return `this - 1` if the expression is const, otherwise call dec(). */
+internal fun IrExpression.decrement(): IrExpression {
+    val thisValue = (this as? IrConst<*>)?.value
+    return when (thisValue) {
```

Comment:
Are byte & short intentionally not handled here? They may appear from const field reads in ConstLowering.

Reply:
These are intentionally not handled here. `decrement()` is only called on expressions which are guaranteed to be one of the 3 progression element types (Int, Long, Char). If you look at UntilHandler, it is called on the argument to `until` *after* at has been casted to the progression element type.